Assertion when using id_print

Sorry that I wasn't able to create a minimum working example, but here's the traceback:
```
Traceback (most recent call last):
  File "demos/encoding.py", line 95, in <module>
    encoding_demo()
  File "demos/encoding.py", line 80, in encoding_demo
    training_trajectory = problem.train()
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/cmm/problem/problem.py", line 31, in train
    trajectory, self.model_parameters, self.iterated_function = train_model(
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/cmm/model/inference.py", line 76, in train_model
    augmented, trajectory = method(model_meta_parameters,
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/api.py", line 169, in f_jitted
    out = xla.xla_call(flat_fun, *args_flat, device=device, backend=backend,
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/core.py", line 1103, in call_bind
    outs = primitive.impl(fun, *args, **params)
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/interpreters/xla.py", line 532, in _xla_call_impl
    compiled_fun = _xla_callable(fun, device, backend, name, donated_invars,
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/linear_util.py", line 221, in memoized_fun
    ans = call(fun, *args)
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/interpreters/xla.py", line 641, in _xla_callable
    out_nodes = jaxpr_subcomp(
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/interpreters/xla.py", line 410, in jaxpr_subcomp
    ans = rule(c, axis_env, extend_name_stack(name_stack, eqn.primitive.name),
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/interpreters/xla.py", line 912, in f
    jaxpr, _, consts = pe.trace_to_jaxpr(
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/interpreters/partial_eval.py", line 429, in trace_to_jaxpr
    jaxpr, (out_pvals, consts, env) = fun.call_wrapped(pvals)
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/linear_util.py", line 150, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/lax/lax_control_flow.py", line 1319, in _scan_impl
    return _scan_impl_loop(
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/lax/lax_control_flow.py", line 1280, in _scan_impl_loop
    _, *outs = while_loop(cond_fun, body_fun, init_val)
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/lax/lax_control_flow.py", line 277, in while_loop
    body_jaxpr, body_consts, body_tree = _initial_style_jaxpr(body_fun, in_tree, init_avals)
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/lax/lax_control_flow.py", line 72, in _initial_style_jaxpr
    jaxpr, out_pvals, consts, out_tree = _initial_style_untyped_jaxpr(
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/lax/lax_control_flow.py", line 66, in _initial_style_untyped_jaxpr
    jaxpr, out_pvals, consts = pe.trace_to_jaxpr(
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/interpreters/partial_eval.py", line 429, in trace_to_jaxpr
    jaxpr, (out_pvals, consts, env) = fun.call_wrapped(pvals)
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/linear_util.py", line 150, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/lax/lax_control_flow.py", line 1270, in body_fun
    out_flat = f_impl(*consts, *carry, *x)
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/core.py", line 142, in jaxpr_as_fun
    return eval_jaxpr(typed_jaxpr.jaxpr, typed_jaxpr.literals, *args)
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/core.py", line 346, in eval_jaxpr
    ans = eqn.primitive.bind(*(subfuns + in_vals), **params)
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/core.py", line 1106, in call_bind
    outs = primitive.process(top_trace, fun, tracers, params)
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/core.py", line 1115, in process
    return trace.process_call(self, fun, tracers, params)
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/interpreters/partial_eval.py", line 232, in process_call
    new_params = update_params(new_params, [not t.pval.is_known() for t in tracers])
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/interpreters/xla.py", line 813, in _xla_call_partial_eval_update_params
    donated_invars = [d for d, uk in zip(donated_invars, in_unknowns) if uk]
  File "/home/neil/.pyenv/versions/3.8.2/lib/python3.8/site-packages/jax/util.py", line 26, in safe_zip
    assert len(arg) == n, 'length mismatch: {}'.format(list(map(len, args)))
AssertionError: length mismatch: [18, 19]
```
If I have time this week, I'll get a MWE.  Hope this helps in the mean time.

@gnecula

--- a/jax/experimental/host_callback.py
+++ b/jax/experimental/host_callback.py
@@ -654,7 +654,10 @@ def _rewrite_eqn(eqn: core.JaxprEqn, eqns: List[core.JaxprEqn],
             dict(
                 eqn.params,
                 call_jaxpr=_rewrite_jaxpr(call_jaxpr, True,
-                                          True)), eqn.source_info))
+                                          True),
+                donated_invars=eqn.params["donated_invars"] + (False,)
+            ),
+          eqn.source_info))
   elif eqn.primitive is custom_derivatives.custom_jvp_call_jaxpr_p:
     fun_jaxpr = eqn.params["fun_jaxpr"]
     new_invars = [*eqn.invars, input_token_var]
@@ -712,7 +715,7 @@ def _rewrite_while_outfeed_cond(eqn: core.JaxprEqn, eqns: List[core.JaxprEqn],
           dict(
               call_jaxpr=transformed_cond_jaxpr.jaxpr,
               name="cond_before",
-              donated_invars=(False,) * (cond_nconsts + len(carry_invars) + 1)),
+              donated_invars=(False,) * len(transformed_cond_jaxpr.in_avals)),
           eqn.source_info))
   # Make a new cond "lambda pred, carry, token: pred"
   new_cond_pred_invar = mk_new_var(cond_jaxpr.out_avals[0])
@@ -751,9 +754,7 @@ def _rewrite_while_outfeed_cond(eqn: core.JaxprEqn, eqns: List[core.JaxprEqn],
           dict(
               call_jaxpr=transformed_body_jaxpr.jaxpr,
               name="body",
-              donated_invars=(False,) *
-              (len(new_body_invars_body_constvars) +
-               len(new_body_invars_carry) + 1 + len(new_body_carry2) + 1)),
+              donated_invars=(False,) * len(transformed_body_jaxpr.in_avals)),
           eqn.source_info),
       core.new_jaxpr_eqn(
           new_body_invars_cond_constvars + new_body_carry2 + [new_body_token2],
@@ -761,8 +762,7 @@ def _rewrite_while_outfeed_cond(eqn: core.JaxprEqn, eqns: List[core.JaxprEqn],
           dict(
               call_jaxpr=transformed_cond_jaxpr.jaxpr,
               name="cond_body",
-              donated_invars=(False,) * (len(new_body_invars_cond_constvars) +
-                                         len(new_body_carry2) + 1 + 2)),
+              donated_invars=(False,) * len(transformed_cond_jaxpr.in_avals)),
           eqn.source_info)
   ]
   new_body_jaxpr = _mk_typed_jaxpr(

--- a/tests/host_callback_test.py
+++ b/tests/host_callback_test.py
@@ -1116,6 +1116,21 @@ what: x times i
           comp, token, 123,
           [xla_bridge.constant(comp, np.zeros((2,), dtype=np.float32))])
 
+  def test_odeint(self):
+    # TODO: find a smaller repro for bug #4015
+    # Seems to be xla_call(scan(xla_call)), all under grad.
+    from jax.experimental.ode import odeint
+
+    def f(x, t, k):
+      x = hcb.id_print(x)
+      return -k * x
+
+    def loss(k=1.0):
+      t = jnp.linspace(0, 0.001, num=2)
+      xs = odeint(f, 1.0, t, k)
+      return xs[-1]
+
+    api.grad(loss)(1.0)  # should not fail
 
 class OutfeedRewriterTest(jtu.JaxTestCase):
 
@@ -1240,7 +1255,7 @@ class OutfeedRewriterTest(jtu.JaxTestCase):
                                                                                  ] b f
                                                                    e = add d 1
                                                                in (c, e, g) }
-                                                  donated_invars=(False, False, False, False, False, False, False)
+                                                  donated_invars=(False, False, False, False)
                                                   name=body ] n p q r
                                 v w = xla_call[ call_jaxpr={ lambda  ; c a b f.
                                                              let _ d g = id_tap[ arg_treedef_=*
@@ -1250,7 +1265,7 @@ class OutfeedRewriterTest(jtu.JaxTestCase):
                                                                                  ] c b f
                                                                  e = lt d 5
                                                              in (e, g) }
-                                                donated_invars=(False, False, False, False, False, False)
+                                                donated_invars=(False, False, False, False)
                                                 name=cond_body ] m s t u
                             in (v, s, t, w) }
                body_nconsts=2

